### PR TITLE
Migrate rustc_monomorphize to use SessionDiagnostic

### PR DESCRIPTION
**Pull request by CleanCut**
### Description

- Migrates diagnostics in `rustc_monomorphize` to use `SessionDiagnostic`
- Adds an `impl IntoDiagnosticArg for PathBuf`

### TODO / Help!
- [x] I'm having trouble figuring out how to apply an optional note. 😕  Help!?
  - Resolved. It was bad docs. Fixed in https://github.com/rust-lang/rustc-dev-guide/pull/1437/files
- [x] `errors:RecursionLimit` should be `#[fatal ...]`, but that doesn't exist so it's `#[error ...]` at the moment.
  - Maybe I can switch after this is merged in? --> https://github.com/rust-lang/rust/pull/100694
  - Or maybe I need to manually implement `SessionDiagnostic` instead of deriving it?
- [x] How does one go about converting an error inside of [a call to struct_span_lint_hir](https://github.com/rust-lang/rust/blob/8064a495086c2e63c0ef77e8e82fe3b9b5dc535f/compiler/rustc_monomorphize/src/collector.rs#L917-L927)?
- [x] ~What placeholder do you use in the fluent template to refer to the value in a vector? It seems like [this code](https://github.com/rust-lang/rust/blob/0b79f758c9aa6646606662a6d623a0752286cd17/compiler/rustc_macros/src/diagnostics/diagnostic_builder.rs#L83-L114) ought to have the answer (or something near it)...but I can't figure it out.~ You can't. Punted.

### CODE REVIEW

File: compiler/rustc_monomorphize/src/errors.rs
Comment (compiler-errors) [inline compiler/rustc_monomorphize/src/errors.rs:19]: `#[note(monomorphize::written_to_path)]` perhaps?

**Comment by JeanCASPAR** [comment]
> I'm having trouble figuring out how to apply an optional note. 😕 Help!?

You can pute the #[note] attribute on a Option<()> field. It will be displayed only when the field is Some, like I did here :
https://github.com/rust-lang/rust/pull/100724/files#diff-5fdce00f0b9b04a72a8bf8d6a0b8866deddf41201e908831c6f7cd444c64bef3R175

**Comment by CleanCut** [comment]
I've taken this PR out of draft status, because I'm now stuck and need a helpful reviewer to guide me forward. 😄

File: compiler/rustc_monomorphize/src/partitioning/mod.rs
Comment (compiler-errors) [inline compiler/rustc_monomorphize/src/partitioning/mod.rs:155]: Why can't you add an `emit_fatal`?

File: compiler/rustc_monomorphize/src/errors.rs
Comment (compiler-errors) [inline compiler/rustc_monomorphize/src/errors.rs:9]: I think `error()` is still correct here -- choosing how to emit the diagnostic (i.e. as an error or as a fatal) is up to the user actually calling `emit_*` on this.

File: compiler/rustc_monomorphize/src/errors.rs
Comment (compiler-errors) [inline compiler/rustc_monomorphize/src/errors.rs:64]: Hm, maybe we should expose some `sess.struct_$SOMETHING` (like `struct_diagnostic`?) that is generic over `EmissionGuarantee`, then make the `SessionDiagnostic` derive generic, i.e.

```
impl<'tcx> SessionDiagnostic for UnusedGenericParams {
  fn into_diagnostic<T: EmissionGuarantee>( .. ) -> DiagnosticBuilder<'tcx, T> {
    let mut diag = sess.struct_diagnostic(rustc_errors:..);
    ..
  }
}
```

cc @davidtwco who might have thoughts

Comment (compiler-errors) [reply compiler/rustc_monomorphize/src/errors.rs:64]: Like, ideally we shouldn't care what kind of `DiagnosticBuilder` (i.e. what kind of emission) we want when designing this struct. We should be able to do `tcx.emit_err(Foo);` or `tcx.emit_warn(Foo);` etc.

File: compiler/rustc_monomorphize/src/errors.rs
Comment (finalchild) [reply compiler/rustc_monomorphize/src/errors.rs:9]: I don't completely understand how diagnostics are handled, but errors and fatals have different type signatures. When errors use ErrorGuaranteed, fatals use !(never). I think that's why existing functions have a separate version for fatals.

My PR(#100694) adds #[fatal(..)], which should be helpful here.

Comment (davidtwco) [reply compiler/rustc_monomorphize/src/errors.rs:9]: This pull request will need to wait for #100694 to land so that it can use `#[fatal(..)]`.

File: compiler/rustc_monomorphize/src/errors.rs
Comment (davidtwco) [reply compiler/rustc_monomorphize/src/errors.rs:64]: I'd be happy with a change like this.

**Comment by davidtwco** [comment]
> errors:RecursionLimit should be #[fatal ...], but that doesn't exist so it's #[error ...] at the moment.

Just skip this one for now and rebase after @finalchild's #100694 lands with `#[fatal(..)]` support.

> How does one go about converting an error inside of a call to `struct_span_lint_hir`?

You can use [`emit_spanned_lint`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.emit_spanned_lint), just use the `LintDiagnostic` derive instead of `SessionDiagnostic` with `#[lint(..)]` instead of `#[error(..)]`. These don't need `#[primary_span]` attributes, which we don't currently prevent but I'm going to make a task about fixing now.

> What placeholder do you use in the fluent template to refer to the value in a vector? It seems like [this code](https://github.com/rust-lang/rust/blob/0b79f758c9aa6646606662a6d623a0752286cd17/compiler/rustc_macros/src/diagnostics/diagnostic_builder.rs#L83-L114) ought to have the answer (or something near it)...but I can't figure it out.

We support `Vec<Span>` fields, if you want to have one message in a bunch of locations, but we otherwise don't have an amazing answer for how to write out lists of items in a translatable way. This is an open question, feel free to read about Fluent and try work out a nice way to do it, then we can follow your approach wherever else this comes up, but you can also just skip a diagnostic with this for now.

File: compiler/rustc_monomorphize/src/errors.rs
Comment (davidtwco) [inline compiler/rustc_monomorphize/src/errors.rs:19]: Does this work if you do `Option<PathBuf>` and then use `was_written` only in the `note` attached to this field? I *think* that might work.

Comment (CleanCut) [reply compiler/rustc_monomorphize/src/errors.rs:19]: I tried that first! Unfortunately, it gives an error that `#[note]` can only be applied to `Span` or `()`. 😞

File: compiler/rustc_monomorphize/src/partitioning/mod.rs
Comment (CleanCut) [reply compiler/rustc_monomorphize/src/partitioning/mod.rs:155]: I don't want to duplicate https://github.com/rust-lang/rust/pull/100694

**Comment by CleanCut** [comment]
> > errors:RecursionLimit should be #[fatal ...], but that doesn't exist so it's #[error ...] at the moment.
> 
> Just skip this one for now and rebase after @finalchild's #100694 lands with `#[fatal(..)]` support.

👍🏻 

> > How does one go about converting an error inside of a call to `struct_span_lint_hir`?
> 
> You can use [`emit_spanned_lint`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.emit_spanned_lint), just use the `LintDiagnostic` derive instead of `SessionDiagnostic` with `#[lint(..)]` instead of `#[error(..)]`. These don't need `#[primary_span]` attributes, which we don't currently prevent but I'm going to make a task about fixing now.

Thanks!!! That's exactly what I needed to know.

✅ Done

---

I've gotten to the point where the code compiles, and I've gotten closer to some tests passing.

I think I'll be able to get most tests passing when I'm able to get the stuff from https://github.com/rust-lang/rust/pull/100694 into this branch. I'm going to wait until it hits `master` just to be safe.

File: compiler/rustc_error_messages/locales/en-US/monomorphize.ftl
Comment (davidtwco) [inline compiler/rustc_error_messages/locales/en-US/monomorphize.ftl:2]:
```suggestion
    reached the recursion limit while instantiating `{$shrunk}`
```

File: compiler/rustc_monomorphize/src/errors.rs
Comment (davidtwco) [inline compiler/rustc_monomorphize/src/errors.rs:67]: I wonder if the suggestion that I've made in https://github.com/rust-lang/rust/pull/100844#discussion_r951399006 would work for this?

File: compiler/rustc_monomorphize/src/partitioning/mod.rs
Comment (davidtwco) [inline compiler/rustc_monomorphize/src/partitioning/mod.rs:155]: This should really be a `UnknownPartitioningStrategy` error that has this message in the Fluent, rather than a `FatalError` type that just has the English string. Likewise with other uses of `{Span,}FatalError`.

**Review by davidtwco — CHANGES_REQUESTED** [review]

**Comment by davidtwco** [comment]
Also, we have a [no merge policy](https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy), so you'll need to remove the merge commits and rebase instead :)

File: compiler/rustc_monomorphize/src/partitioning/mod.rs
Comment (CleanCut) [reply compiler/rustc_monomorphize/src/partitioning/mod.rs:155]: Will do 👍🏻

**Comment by CleanCut** [comment]
> Also, we have a [no merge policy](https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy), so you'll need to remove the merge commits and rebase instead :)

Ah, TIL. Will fix. 👍🏻

File: compiler/rustc_error_messages/locales/en-US/monomorphize.ftl
Comment (CleanCut) [reply compiler/rustc_error_messages/locales/en-US/monomorphize.ftl:2]: Ha! I fixed this and forgot to push it up.

**Comment by CleanCut** [comment]
Rebased. Working on the other items now.

File: compiler/rustc_monomorphize/src/partitioning/mod.rs
Comment (CleanCut) [reply compiler/rustc_monomorphize/src/partitioning/mod.rs]: Done with `UnknownPartitionStrategy` and `SymbolAlreadyDefined`.

**Comment by CleanCut** [comment]
There is currently one test failing for switching the order of a `note:` and a `help:` line. I have tried rearranging the struct, but that doesn't seem to affect anything. Is this sort of change okay to bless?

```
---- [ui] src/test/ui/type_length_limit.rs stdout ----
diff of stderr:

4       LL | pub fn drop<T>(_x: T) {}
5          | ^^^^^^^^^^^^^^^^^^^^^
6          |
-          = note: the full type name has been written to '$TEST_BUILD_DIR/type_length_limit/type_length_limit.long-type.txt'
8          = help: consider adding a `#![type_length_limit="8"]` attribute to your crate
+          = note: the full type name has been written to '$TEST_BUILD_DIR/type_length_limit/type_length_limit.long-type.txt'
9
10      error: reached the type-length limit while instantiating `<[closure@std::rt::lang_start<()...e<()>>::call_once - shim(vtable)`
11        --> $SRC_DIR/core/src/ops/function.rs:LL:COL

13      LL |     extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
14         |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
15         |
-          = note: the full type name has been written to '$TEST_BUILD_DIR/type_length_limit/type_length_limit.long-type.txt'
17         = help: consider adding a `#![type_length_limit="8"]` attribute to your crate
+          = note: the full type name has been written to '$TEST_BUILD_DIR/type_length_limit/type_length_limit.long-type.txt'
18
19      error: aborting due to 2 previous errors
20
```

**Comment by davidtwco** [comment]
> There is currently one test failing for switching the order of a `note:` and a `help:` line. I have tried rearranging the struct, but that doesn't seem to affect anything. Is this sort of change okay to bless?

In this instance, it probably would be, yeah. I find this curious though, changing the order of the attributes should influence their order on the diagnostic. Looking at the struct, I guess it is because both are spanless subdiagnostics and one is defined on the fields and the other on the struct, so we can't just swap the two lines.

File: compiler/rustc_monomorphize/src/collector.rs
Comment (davidtwco) [reply compiler/rustc_monomorphize/src/collector.rs:1334]: For now, just change `require` to return some appropriate error type which isn't the English error message, e.g. `struct LangItemError(LangItem)`. Add a function to that new type that returns the current `String` - use that everywhere else to avoid having to make changes to lots of other use sites. Implement `IntoDiagnosticArg` for it so that you can just use it in the `SessionDiagnostic` in this crate.

Comment (CleanCut) [reply compiler/rustc_monomorphize/src/collector.rs:1334]: Excellent idea!

✅ Done

**Comment by CleanCut** [comment]
> > There is currently one test failing for switching the order of a `note:` and a `help:` line. I have tried rearranging the struct, but that doesn't seem to affect anything. Is this sort of change okay to bless?
> 
> In this instance, it probably would be, yeah. I find this curious though, changing the order of the attributes should influence their order on the diagnostic. Looking at the struct, I guess it is because both are spanless subdiagnostics and one is defined on the fields and the other on the struct, so we can't just swap the two lines.

✅ Blessed.

**Comment by CleanCut** [comment]
😢  I'm getting some codegen failures on Apple Silicon. No idea why.

<details><summary>aarch64-apple-darwin codegen error output</summary>

```
Check compiletest suite=codegen mode=codegen (aarch64-apple-darwin -> aarch64-apple-darwin)

running 350 tests
iiiiiiiiii...ii......i................iii.iiiii.....i........i...................i...... 88/350
........ii..iii.......i...................i............i........i....i......ii......i..i 176/350
..................i...i.i.i....iiii..iii....iiiiiii.iiiiiii....F....................i... 264/350
..i......ii......iii.......i.........................ii.....i.i.......................
failures:

---- [codegen] src/test/codegen/mem-replace-direct-memcpy.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/Users/nathan/rust/rust/build/aarch64-apple-darwin/ci-llvm/bin/FileCheck" "--input-file" "/Users/nathan/rust/rust/build/aarch64-apple-darwin/test/codegen/mem-replace-direct-memcpy/mem-replace-direct-memcpy.ll" "/Users/nathan/rust/rust/src/test/codegen/mem-replace-direct-memcpy.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC"
stdout: none
--- stderr -------------------------------
/Users/nathan/rust/rust/src/test/codegen/mem-replace-direct-memcpy.rs:20:11: error: CHECK: expected string not found in input
// CHECK: call void @llvm.memcpy.{{.+}}({{i8\*|ptr}} align 1 %{{.*}}, {{i8\*|ptr}} align 1 %dest, i{{.*}} 1, i1 false)
          ^
/Users/nathan/rust/rust/build/aarch64-apple-darwin/test/codegen/mem-replace-direct-memcpy/mem-replace-direct-memcpy.ll:47:21: note: scanning from here
; core::mem::replace
                    ^
/Users/nathan/rust/rust/build/aarch64-apple-darwin/test/codegen/mem-replace-direct-memcpy/mem-replace-direct-memcpy.ll:129:2: note: possible intended match here
 call void @llvm.memcpy.p0.p0.i64(ptr align 1 %_4, ptr align 1 %src, i64 1, i1 false)
 ^

Input file: /Users/nathan/rust/rust/build/aarch64-apple-darwin/test/codegen/mem-replace-direct-memcpy/mem-replace-direct-memcpy.ll
Check file: /Users/nathan/rust/rust/src/test/codegen/mem-replace-direct-memcpy.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            .
            .
            .
           42: define internal i8 @"_ZN4core3mem13manually_drop21ManuallyDrop$LT$T$GT$10into_inner17h98f1424f9d2c8ff5E"(i8 %slot) unnamed_addr #0 { 
           43: start: 
           44:  ret i8 %slot 
           45: } 
           46:  
           47: ; core::mem::replace 
check:20'0                         X error: no match found
           48: ; Function Attrs: inlinehint uwtable 
check:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49: define internal i8 @_ZN4core3mem7replace17hd07fab279ae0e11dE(ptr noalias noundef align 1 dereferenceable(1) %dest, i8 %src) unnamed_addr #1 personality ptr @rust_eh_personality { 
check:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50: start: 
check:20'0     ~~~~~~~
           51:  %0 = alloca { ptr, i32 }, align 8 
check:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52:  %_7 = alloca i8, align 1 
check:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
            .
          124: ; call core::mem::maybe_uninit::MaybeUninit<T>::as_mut_ptr 
check:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          125:  %_4 = call ptr @"_ZN4core3mem12maybe_uninit20MaybeUninit$LT$T$GT$10as_mut_ptr17h25c1c862ec6279c9E"(ptr noalias noundef align 1 dereferenceable(1) %tmp) 
check:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          126:  br label %bb2 
check:20'0     ~~~~~~~~~~~~~~~
          127:  
check:20'0     ~
          128: bb2: ; preds = %bb1 
check:20'0     ~~~~~~~~~~~~~~~~~~~~
          129:  call void @llvm.memcpy.p0.p0.i64(ptr align 1 %_4, ptr align 1 %src, i64 1, i1 false) 
check:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:20'1      ?                                                                                     possible intended match
          130:  %_6 = load i8, ptr %tmp, align 1 
check:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          131: ; call core::mem::maybe_uninit::MaybeUninit<T>::assume_init 
check:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          132:  %1 = call i8 @"_ZN4core3mem12maybe_uninit20MaybeUninit$LT$T$GT$11assume_init17hbd6e80e342715b18E"(i8 %_6, ptr noalias noundef readonly align 8 dereferenceable(24) @alloc21) 
check:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          133:  br label %bb3 
check:20'0     ~~~~~~~~~~~~~~~
          134:  
check:20'0     ~
            .
            .
            .
>>>>>>
------------------------------------------



failures:
    [codegen] src/test/codegen/mem-replace-direct-memcpy.rs

test result: FAILED. 274 passed; 1 failed; 75 ignored; 0 measured; 0 filtered out; finished in 2.25s

Some tests failed in compiletest suite=codegen mode=codegen host=aarch64-apple-darwin target=aarch64-apple-darwin
```
</details>

Other than that, this PR is pretty much there.

File: compiler/rustc_monomorphize/src/errors.rs
Comment (CleanCut) [reply compiler/rustc_monomorphize/src/errors.rs:58]: I don't think I have the bandwidth to take on that change at the moment. Will it block this PR?

**Comment by CleanCut** [comment]
✅ CI Passing, ready for re-review

⚠️ I'm still experiencing [a codegen failure locally on Apple Silicon](https://github.com/rust-lang/rust/pull/100730#issuecomment-1226528891).

File: compiler/rustc_monomorphize/src/errors.rs
Comment (davidtwco) [reply compiler/rustc_monomorphize/src/errors.rs:58]: We can loop back for this later.

Comment (davidtwco) [reply compiler/rustc_monomorphize/src/errors.rs:58]: Is this not triggering the lint though?

**Comment by davidtwco** [comment]
> ✅ CI Passing, ready for re-review
> 
> ⚠️ I'm still experiencing [a codegen failure locally on Apple Silicon](https://github.com/rust-lang/rust/pull/100730#issuecomment-1226528891).

I assume that the codegen failure is spurious because nothing you've changed should make a difference there - once you've rebased and fixed the conflicts, we can try merging this. Thanks for sticking with it :)

**Comment by CleanCut** [comment]
> I assume that the codegen failure is spurious because nothing you've changed should make a difference there - once you've rebased and fixed the conflicts, we can try merging this. Thanks for sticking with it :)

I hope so. The local codegen  failure persists even after the rebase.

File: compiler/rustc_monomorphize/src/polymorphize.rs
Comment (davidtwco) [inline compiler/rustc_monomorphize/src/polymorphize.rs:219]: Can you remove this comment?

**Review by davidtwco — COMMENTED** [review]
I think we're almost ready to land this - you've pushed a submodule update for Cargo though, which shouldn't be there.

File: compiler/rustc_monomorphize/src/polymorphize.rs
Comment (CleanCut) [reply compiler/rustc_monomorphize/src/polymorphize.rs:219]: ✅

**Comment by CleanCut** [comment]
> I think we're almost ready to land this - you've pushed a submodule update for Cargo though, which shouldn't be there.

✅ Fixed the errant submodule change. Ready for re-review.

File: compiler/rustc_monomorphize/src/polymorphize.rs
Comment (CleanCut) [reply compiler/rustc_monomorphize/src/polymorphize.rs]: Forgot to note that this has been done.

**Review by davidtwco — APPROVED** [review]

**Comment by klensy** [comment]
Submodule changes still in commits history, can you actually rebase?

**Pull request merged**
